solvers/DPGD.py

```diff
@@ -34,19 +34,27 @@ def set_objective(self, A, reg, y, c, delta, data_fit):
 
     def run(self, callback):
         n, p = self.A.shape
-        L = np.tri(p)[:, 1:]
-        DA_inv = np.linalg.pinv(self.A @ L)
+        if isinstance(self.A, np.ndarray):
+            DA_inv = np.diff(np.linalg.pinv(self.A), axis=0)
+            DA_invDA_invt = DA_inv @ DA_inv.T
+            DA_invy = DA_inv @ self.y
+            AtA_inv = np.linalg.pinv(self.A.T @ self.A)
+        else:
+            D_inv = np.linspace(1/p - 1, - 1/p, p-1) + np.tri(p)[:, 1:]
+            DA_inv = np.linalg.pinv(self.A @ D_inv)
```

Comment:
Did you check this gives the right inverse? If so, could you add a test in `tests/test_matrices_operations.py`?

Reply:
A simple check: you should get the same results, whether you pass `A` as a ndarray or wrapped in a `LinearOperator`.
Can you check this?

Reply:
There is a problem here. I will fix it.

Reply:
I have check that it's the right inverse, however it works only when A is ```LinearOperator```.

Reply:
how did you check this is the right inverse?

Reply:
'''python
def test_inverse(D, D_inv):
    n = D.shape[0]
    return np.allclose(D @ D_inv, np.identity(n))
```